CircleSpawn/CircleSpawn/CircleSpawnController.swift

```diff
@@ -1,11 +1,104 @@
 import UIKit
 
-class CircleSpawnController: UIViewController {
-
-	// TODO: Assignment 1
-
+class SpawnedViewSpawnController: UIViewController, UIGestureRecognizerDelegate {
+    
+    
+    private let size: CGFloat = 100
+    
+    //TODO: Move offsets to CircleView
+    var xOffset: CGFloat = 0.0
+    var yOffset: CGFloat = 0.0
+    
 	override func loadView() {
 		view = UIView()
 		view.backgroundColor = .white
 	}
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        let doubleTap = UITapGestureRecognizer(target: self, action: #selector(handleDoubleTap(_:)))
+        doubleTap.numberOfTapsRequired = 2
+        doubleTap.delegate = self
+        view.addGestureRecognizer(doubleTap)
+    }
+    
+    @objc func handleDoubleTap(_ tap: UITapGestureRecognizer) {
+        let spawnedView = CircleView(color: UIColor.randomBrightColor(), center: tap.location(in: view))
+        view.addSubview(spawnedView)
+        
+        spawnedView.alpha = 0
+        spawnedView.transform = CGAffineTransform(scaleX: 1.5, y: 1.5)
+        UIView.animate(withDuration: 0.2, animations: {
+            spawnedView.alpha = 1
+            spawnedView.transform = .identity
+        }, completion: { completed in })
+        
+        let tripleTap = UITapGestureRecognizer(target: self, action: #selector(handleTripleTap(_:)))
+        tripleTap.numberOfTapsRequired = 3
+        spawnedView.addGestureRecognizer(tripleTap)
+        tap.require(toFail: tripleTap)
+        
+        let longPress = UILongPressGestureRecognizer(target: self, action: #selector(handleLongPress(_:)))
+        spawnedView.addGestureRecognizer(longPress)
+    }
+    
+    @objc func handleTripleTap(_ tap: UITapGestureRecognizer) {
+        UIView.animate(withDuration: 0.2, animations: {
+            tap.view?.alpha = 0
+            tap.view?.transform = CGAffineTransform(scaleX: 1.5, y: 1.5)
+        }, completion: { completed in
+            tap.view?.removeFromSuperview()
+        })
+    }
+    
+    @objc func handleLongPress(_ longPress: UILongPressGestureRecognizer) {
+        
+            switch longPress.state {
+            case .began:
+                UIView.animate(withDuration: 0.2, animations: {
+                    longPress.view?.alpha = 0.5
+                    longPress.view?.transform = CGAffineTransform(scaleX: 1.5, y: 1.5)
+                }, completion: { completed in })
+                
+                xOffset = (longPress.view?.center.x)! - longPress.location(in: view).x
+                yOffset = (longPress.view?.center.y)! - longPress.location(in: view).y
```

Comment:
To są obliczenia wykonywane wielokrotnie. Nie wiesz ile trwa policzenie `longPress.location(in view: UIView)`. Lepiej policzyć to raz i skorzystać z wartości.

Reply:
Jak nie w handleLongPress to gdzie muszę liczyć offsety? To się wykonuje tylko w .began, jednokrotnie. Po offsetach napisałem print(xOffset) i drukowało tylko jeden raz jak łapałem kółko.